text/0000-interior.md

```diff
@@ -0,0 +1,348 @@
+fn as_interior<T>(t: &T) -> &Interior<T>;
+fn as_interior_mut<T>(t: &mut T) -> &mut Interior<T>;
+fn of_interior<T>(t: &Interior<T>) -> &T;
+fn of_interior_mut<T>(t: &mut Interior<T>) -> &mut T;
+```
+
+And then it is possible to define a compatibility `Drop` trait in
+terms of `DropValue`:
+
+```rust
+impl<T: Drop> DropValue for T {
+  fn drop_value(this: Interior<Self>) {
+    of_interior_mut(this).drop();
+  }
+}
+```
+
+## Compiler-generated drop-glue.
+
+In this design, the compiler-generated drop-glue for a type gets so
+simplified as to become nearly, or possibly entirely, non-existent.
+Consider the following types:
+
+```rust
+struct Foo;
+impl Drop for Foo { ... }
+
+struct Bar(Foo, Foo);
+impl Drop for Bar { ... }
+
+enum Baz {
+  Empty,
+  AmFoo(Foo),
+  AmBar(Bar),
+}
+impl Drop for Baz { ... }
+```
+
+In current Rust, the drop-glue for these types will look something
+like this:
+
+```rust
+fn drop_glue_Foo(v: Foo) {
+  v.drop();
+  mem::forget(v);
+}
+fn drop_glue_Bar(v: Bar) {
+  v.drop();
+  drop_glue_Foo(v.0);
+  drop_glue_Foo(v.1);
+  mem::forget(v);
+}
+fn drop_glue_Baz(v: Baz) {
+  v.drop();
+  match v {
+    Empty => (),
+    AmFoo(val) => drop_glue_Foo(val),
+    AmBar(val) => drop_glue_Bar(val),
+  }
+  mem::forget(v);
+}
+```
+
+Under this proposal, the drop-glue for member fields is called at
+the point `this` is consumed:
+
+```rust
+impl DropValue for Foo {
+  fn drop_value(this: Interior<Foo>) {
+    println!("drop_foo");
+    // no compiler-inserted clean-up necessary
+  }
+}
+impl DropValue for Bar {
+  fn drop_value(this: Interior<Bar>) {
+    println!("drop_bar");
+    // "this" is consumed at function exit, so the compiler
+    // will generate the following code:
+    // DropValue::<Foo>::drop_value(into_interior(this.0));
+    // DropValue::<Foo>::drop_value(into_interior(this.1));
+    // and the following will be displayed as this function is called:
+    // drop_bar
+    // drop_foo
+    // drop_foo
+  }
+}
+impl DropValue for Baz {
+  fn drop_value(this: Interior<Baz>) {
+    // mem::drop() is #[inline], so the compiler will replace this
+    // call:
+    mem::drop(this);
+    // with something like the following code:
+    // match this.discriminant {
+    //   Empty => (),
+    //   AmFoo(val) => DropValue::<Foo>::drop_value(into_interior(val)),
+    //   AmBar(val) => DropValue::<Bar>::drop_value(into_interior(val)),
+    // }
+    println!("drop_baz");
+    // note that the println! should be evaluated *after* `this` is
+    // dropped. So if `this` were `AmFoo(Foo)`, then the following
+    // would be displayed:
+    // drop_foo
+    // drop_baz
+  }
+}
+```
+
+The drop-glue function no longer appears to be necessary: all clean-up
+now occurs in the drop-hook itself, in the same way it does for types
+that do not have drop-glue defined.
+
+## `impl DropValue for Interior<T>`
+
+We do not prohibit `impl DropValue for Interior<T>`: it should just
+work in the natural way.
```

Comment:
Well, I had thought `DropValue` would be defined in `libcore`, same as `Drop`. So I think this should be valid per orphan rules.